extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultConfigSource.java

```diff
@@ -0,0 +1,264 @@
+    }
+
+    // ---
+
+    private VaultManager getVaultManager() {
+
+        VaultServerRuntimeConfig serverConfig = getConfig();
+
+        // init at most once
+        if (init.compareAndSet(false, true)) {
+            VaultManager.init(serverConfig);
+        }
+
+        return VaultManager.getInstance();
+    }
+
+    private VaultServerRuntimeConfig getConfig() {
+        VaultServerRuntimeConfig serverConfig = this.serverConfig.get();
+        if (serverConfig != null) {
+            return serverConfig;
+        } else {
+            serverConfig = loadConfig();
+            log.debug("loaded vault server config " + serverConfig);
+            this.serverConfig.set(serverConfig);
+            return this.serverConfig.get();
+        }
+    }
+
+    // need to recode config loading since we are at the config source level
+    private VaultServerRuntimeConfig loadConfig() {
+
+        VaultServerRuntimeConfig serverConfig = new VaultServerRuntimeConfig();
+        serverConfig.url = newURL(getOptionalVaultProperty("url"));
+        serverConfig.authenticationType = VaultAuthenticationType
+                .valueOf(getVaultProperty("authentication-type", KUBERNETES.name()).toUpperCase());
+        serverConfig.kubernetesAuthenticationRole = getOptionalVaultProperty("kubernetes-authentication-role");
+        serverConfig.kubernetesJwtTokenPath = getVaultProperty("kubernetes-jwt-token-path", DEFAULT_KUBERNETES_JWT_TOKEN_PATH);
+        serverConfig.userpassUsername = getVaultProperty("userpass-username");
+        serverConfig.userpassPassword = getVaultProperty("userpass-password");
+        serverConfig.vaultToken = getOptionalVaultProperty("vault-token");
+        serverConfig.renewGracePeriod = getVaultDuration("renew-grace-period", DEFAULT_RENEW_GRACE_PERIOD);
+        serverConfig.secretConfigCachePeriod = getVaultDuration("secret-config-cache-period",
+                DEFAULT_SECRET_CONFIG_CACHE_PERIOD);
+        serverConfig.logConfidentialityLevel = LogConfidentialityLevel
+                .valueOf(getVaultProperty("log-confidentiality-level", MEDIUM.name()).toUpperCase());
+        serverConfig.kvSecretEngineVersion = parseInt(
+                getVaultProperty("kv-secret-engine-version", KV_SECRET_ENGINE_VERSION_V1));
+        serverConfig.kvSecretEngineMountPath = getVaultProperty("kv-secret-engine-mount-path",
+                DEFAULT_KV_SECRET_ENGINE_MOUNT_PATH);
+        serverConfig.secretConfigKvPath = getOptionalVaultProperty("secret-config-kv-path");
+        serverConfig.tlsSkipVerify = parseBoolean(getVaultProperty("tls-skip-verify", DEFAULT_TLS_SKIP_VERIFY));
+        serverConfig.tlsCaCert = getOptionalVaultProperty("tls-ca-cert");
+        serverConfig.connectTimeout = getVaultDuration("connect-timeout", DEFAULT_CONNECT_TIMEOUT);
+        serverConfig.readTimeout = getVaultDuration("read-timeout", DEFAULT_READ_TIMEOUT);
```

Comment:
I'm not sure to understand why you need to recode the config object construction? I might miss something obvious though.

Reply:
see https://github.com/quarkusio/quarkus/pull/2897/#discussion_r326099504